libclc/clc/lib/generic/math/clc_log_base.h

```diff
@@ -0,0 +1,339 @@
+__clc_log2(float x)
+#elif defined(COMPILING_LOG10)
+__clc_log10(float x)
+#else
+__clc_log(float x)
+#endif
+{
+
+#if defined(COMPILING_LOG2)
+  const float LOG2E = 0x1.715476p+0f;      // 1.4426950408889634
+  const float LOG2E_HEAD = 0x1.700000p+0f; // 1.4375
+  const float LOG2E_TAIL = 0x1.547652p-8f; // 0.00519504072
+#elif defined(COMPILING_LOG10)
+  const float LOG10E = 0x1.bcb7b2p-2f;        // 0.43429448190325182
+  const float LOG10E_HEAD = 0x1.bc0000p-2f;   // 0.43359375
+  const float LOG10E_TAIL = 0x1.6f62a4p-11f;  // 0.0007007319
+  const float LOG10_2_HEAD = 0x1.340000p-2f;  // 0.30078125
+  const float LOG10_2_TAIL = 0x1.04d426p-12f; // 0.000248745637
+#else
+  const float LOG2_HEAD = 0x1.62e000p-1f;  // 0.693115234
+  const float LOG2_TAIL = 0x1.0bfbe8p-15f; // 0.0000319461833
+#endif
+
+  uint xi = __clc_as_uint(x);
+  uint ax = xi & EXSIGNBIT_SP32;
+
+  // Calculations for |x-1| < 2^-4
+  float r = x - 1.0f;
+  int near1 = __clc_fabs(r) < 0x1.0p-4f;
+  float u2 = MATH_DIVIDE(r, 2.0f + r);
+  float corr = u2 * r;
+  float u = u2 + u2;
+  float v = u * u;
+  float znear1, z1, z2;
+
+  // 2/(5 * 2^5), 2/(3 * 2^3)
+  z2 = __clc_mad(u, __clc_mad(v, 0x1.99999ap-7f, 0x1.555556p-4f) * v, -corr);
+
+#if defined(COMPILING_LOG2)
+  z1 = __clc_as_float(__clc_as_int(r) & 0xffff0000);
+  z2 = z2 + (r - z1);
+  znear1 = __clc_mad(
+      z1, LOG2E_HEAD,
+      __clc_mad(z2, LOG2E_HEAD, __clc_mad(z1, LOG2E_TAIL, z2 * LOG2E_TAIL)));
+#elif defined(COMPILING_LOG10)
+  z1 = __clc_as_float(__clc_as_int(r) & 0xffff0000);
+  z2 = z2 + (r - z1);
+  znear1 = __clc_mad(
+      z1, LOG10E_HEAD,
+      __clc_mad(z2, LOG10E_HEAD, __clc_mad(z1, LOG10E_TAIL, z2 * LOG10E_TAIL)));
+#else
+  znear1 = z2 + r;
+#endif
+
+  // Calculations for x not near 1
+  int m = (int)(xi >> EXPSHIFTBITS_SP32) - EXPBIAS_SP32;
+
+  // Normalize subnormal
+  uint xis = __clc_as_uint(__clc_as_float(xi | 0x3f800000) - 1.0f);
+  int ms = (int)(xis >> EXPSHIFTBITS_SP32) - 253;
+  int c = m == -127;
+  m = c ? ms : m;
+  uint xin = c ? xis : xi;
+
+  float mf = (float)m;
+  uint indx = (xin & 0x007f0000) + ((xin & 0x00008000) << 1);
+
+  // F - Y
+  float f = __clc_as_float(0x3f000000 | indx) -
+            __clc_as_float(0x3f000000 | (xin & MANTBITS_SP32));
+
+  indx = indx >> 16;
+  r = f * USE_TABLE(log_inv_tbl, indx);
+
+  // 1/3,  1/2
+  float poly = __clc_mad(__clc_mad(r, 0x1.555556p-2f, 0.5f), r * r, r);
+
+#if defined(COMPILING_LOG2)
+  float2 tv = USE_TABLE(log2_tbl, indx);
+  z1 = tv.s0 + mf;
+  z2 = __clc_mad(poly, -LOG2E, tv.s1);
+#elif defined(COMPILING_LOG10)
+  float2 tv = USE_TABLE(log10_tbl, indx);
+  z1 = __clc_mad(mf, LOG10_2_HEAD, tv.s0);
+  z2 = __clc_mad(poly, -LOG10E, mf * LOG10_2_TAIL) + tv.s1;
+#else
+  float2 tv = USE_TABLE(log_tbl, indx);
+  z1 = __clc_mad(mf, LOG2_HEAD, tv.s0);
+  z2 = __clc_mad(mf, LOG2_TAIL, -poly) + tv.s1;
+#endif
+
+  float z = z1 + z2;
+  z = near1 ? znear1 : z;
+
+  // Corner cases
+  z = ax >= PINFBITPATT_SP32 ? x : z;
```

Comment:
These types of exponent checks should be replaced with proper isnan/isinf/isfinite checks